uniffi_bindgen/src/interface/types/dispatch.rs

```diff
@@ -0,0 +1,196 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Implement type behavior using the dispatch pattern
+//!
+//! There are a couple issues with implementing type behavior:
+//!
+//!    - `Type` has a lot of variants.  It can be hard to implement traits on `Type` when each
+//!    function needs to match against them all.
+//!    - A secondary issue is that there are types that correspond to `Type` variants (Record,
+//!      Object, CallbackInterface, etc.)  It would be nice for these types to share behavior with
+//!      their `Type` variant.
+//!
+//! This module helps solve both issues using the dispatch pattern and macros:
+//!
+//!  - Define a set of handler structs.  Each one will handle some subset of `Type` variants
+//!    (usually 1).
+//!  - Define a macro that can dispatch function calls for `Type` to one of the handler structs.
+//!  - Wrap traits with the `type_dispatch!` macro, defined here.  `type_dispatch!` derives a trait
+//!    impl for `Type`, `Record`, `Object`, `CallbackInterface`, etc. by dispatching the calls to
+//!    the appropriate handler type.  See `CodeType` and `KotlinCodeType` for an example.
```

Comment:
Consider using the ``[`CodeType`]`` syntax here, so that when this doc comment gets rendered into the online docs, the reader can click a link to jump directly to the examples in question.